### PR TITLE
sync: Tier selector

### PR DESCRIPTION
**Pull request by aditya-nambiar**

### CODE REVIEW

File: fennel/lib/includes/include_mod.py [hunk 1]
```diff
@@ -20,3 +22,65 @@ def decorator(func: Callable) -> Callable:
...
+    TierSelector is a mixin class that can be used to specify the tiers the entity supports.
+    """
+
+    tiers: Optional[str | List[str]] = None
+
+    def __init__(self, tiers):
+        super().__init__(tiers=tiers)
+
+    @validator("tiers", pre=True, each_item=True)
+    def check_string(cls, v):
```
Comment (nikhilgarg28) [inline fennel/lib/includes/include_mod.py:38]: I assume "~" is not a valid tier string either?

File: fennel/sources/sources.py [hunk 2]
```diff
@@ -366,6 +365,7 @@ class DataConnector:
     lateness: Duration
     cdc: str
     starting_from: Optional[datetime] = None
+    tiers: TierSelector
```
Comment (nikhilgarg28) [inline fennel/sources/sources.py:368]: Will we expect most usages of this to be a single string or a list? I suspect it's a single list but what do you think? If you also agree, I think this argument should be called "tier" (otherwise `tiers='prod'`) reads weird. 

Same for `tiers` everywhere (extractors, pipelines, sources etc)

File: fennel/sources/test_sources.py [hunk 3]
```diff
@@ -224,6 +225,114 @@ class UserInfoDataset:
 )
 
 
+def test_tier_selector_on_source():
```
Comment (nikhilgarg28) [inline fennel/sources/test_sources.py:228]: Worth adding a test where multiple sources/pipelines/extractors exist but none get selected in the end?

File: fennel/client_tests/test_tier_selector.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        uniq = sale.groupby("ticket_id").first()
+        c = (
+            uniq.join(info, how="inner", on=["title"])
+            .explode(columns=["actors"])
+            .rename(columns={"actors": "name"})
+        )
+        # name -> Option[str]
+        schema = c.schema()
+        schema["name"] = str
+        c = c.transform(lambda x: x, schema)
```
Comment (nikhilgarg28) [inline fennel/client_tests/test_tier_selector.py:60]: Maybe use `dropnull`?

File: fennel/client_tests/test_tier_selector.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        uniq = sale.groupby("ticket_id").first()
+        c = (
+            uniq.join(info, how="inner", on=["title"])
+            .explode(columns=["actors"])
+            .rename(columns={"actors": "name"})
+        )
+        # name -> Option[str]
+        schema = c.schema()
+        schema["name"] = str
+        c = c.transform(foo, schema)
```
Comment (nikhilgarg28) [inline fennel/client_tests/test_tier_selector.py:87]: Same here?

**Review by nikhilgarg28 — APPROVED** [review]

File: docs/examples/getting-started/quickstart.py [hunk 6]
```diff
@@ -23,14 +23,20 @@
...
 
 # /docsnip
 
 
 # docsnip datasets
 @dataset
-@source(postgres.table("product_info", cursor="last_modified"), every="1m")
+@source(
+    postgres.table("product_info", cursor="last_modified"),
+    every="1m",
```
Comment (nikhilgarg28) [inline docs/examples/getting-started/quickstart.py:36]: can you please remove every field so that we can hopefully still fit the `source` decorator in a single line? If that doesn't work, try shortening the postgres table name from "product_info" to just "product". 

This is the first line of code shown in demo so I really want it to feel neat/simple.

File: docs/examples/overview/concepts.py [hunk 7]
```diff
@@ -26,11 +26,15 @@ class UserDataset:
 
 postgres = Postgres.get(name="postgres")
 kafka = Kafka.get(name="kafka")
+webhook = Webhook(name="fennel_webhook")
 
 
 # docsnip external_data_sources
 @meta(owner="data-eng-oncall@fennel.ai")
-@source(postgres.table("user", cursor="update_timestamp"), every="1m")
+@source(
```
Comment (nikhilgarg28) [inline docs/examples/overview/concepts.py:34]: Can you try fitting this in one line somehow? e.g. rename postgres to  "pg" or change the name of cursor field?

**Review by nikhilgarg28 — COMMENTED** [review]
Requesting couple quick cosmetic changes to maintain cleanliness of demo & high traffic parts of the docs.
